File: leetcode-daily/april-2024/678-valid-parenthesis-string.cpp
/*
https://leetcode.com/problems/valid-parenthesis-string

Given a string `s` containing only three types of characters: `'('`, `')'` and `'*'`, return `true`
*if* `s` *is
**valid***.

The following rules define a **valid** string:

* Any left parenthesis `'('` must have a corresponding right parenthesis `')'`.
* Any right parenthesis `')'` must have a corresponding left parenthesis `'('`.
* Left parenthesis `'('` must go before the corresponding right parenthesis `')'`.
* `'*'` could be treated as a single right parenthesis `')'` or a single left parenthesis `'('` or
an empty string `""`.


**Example 1:**

**Input:** s = "()"
**Output:** true

**Example 2:**

**Input:** s = "(*)"
**Output:** true

**Example 3:**

**Input:** s = "(*))"
**Output:** true


**Constraints:**

* `1 <= s.length <= 100`
* `s[i]` is `'('`, `')'` or `'*'`.
*/

#include "./../../cpp-utils/printutils.h"
#include "bits/stdc++.h"
#include <iostream>
using namespace std;

class Solution {
  public:
    bool checkValidString(string s) {
        stack<pair<char, int>> left;
        queue<pair<char, int>> right;
        deque<pair<char, int>> astr;
        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];
            if (c == '(') {
                left.push({c, i});
            } else if (c == ')') {
                if (left.empty()) {
                    right.push({c, i});
                } else {
                    left.pop();
                }
            } else {
                astr.push_back({c, i});
            }
        }

        while (!left.empty() && !astr.empty()) {
            if (left.top().second < astr.front().second) {
                left.pop();
                astr.pop_front();
            } else if (left.top().second < astr.back().second) {
                left.pop();
                astr.pop_back();
            } else {
                break;
            }
        }

        while (!right.empty() && !astr.empty()) {
            if (right.front().second > astr.back().second) {
                right.pop();
                astr.pop_back();
            } else if (right.front().second > astr.front().second) {
                right.pop();
                astr.pop_front();
            } else {
                break;
            }
        }

        return left.empty() && right.empty();
    }
};

int main() {
    Solution sol;

    string s_1 = "()";
    bool expected_1 = true;
    bool output_1 = sol.checkValidString(s_1);
    cout << output_1 << endl;
    assert(output_1 == expected_1);

    string s_2 = "(*)";
    bool expected_2 = true;
    bool output_2 = sol.checkValidString(s_2);
    cout << output_2 << endl;
    assert(output_2 == expected_2);

    string s_3 = "(*))";
    bool expected_3 = true;
    bool output_3 = sol.checkValidString(s_3);
    cout << output_3 << endl;
    assert(output_3 == expected_3);

    string s_4 = "(((((*(()((((*((**(((()()*)()()()*((((**)())*)*)))))))(())(()))())((*()()(((()((("
                 ")*(())*(()**)()(())";
    bool expected_4 = true;
    bool output_4 = sol.checkValidString(s_4);
    cout << output_4 << endl;
    assert(output_4 == expected_4);

    string s_5 = "((((()(()()()*()(((((*)()*(**(())))))(())()())(((())())())))))))(((((())*)))()))("
                 "()((*()*(*)))(*)()";
    bool expected_5 = true;
    bool output_5 = sol.checkValidString(s_5);
    cout << output_5 << endl;
    assert(output_5 == expected_5);

    return 0;
}

File: leetcode-daily/april-2024/678-valid-parenthesis-string.rs
/*
https://leetcode.com/problems/valid-parenthesis-string

Given a string `s` containing only three types of characters: `'('`, `')'` and `'*'`, return `true` *if* `s` *is
**valid***.

The following rules define a **valid** string:

* Any left parenthesis `'('` must have a corresponding right parenthesis `')'`.
* Any right parenthesis `')'` must have a corresponding left parenthesis `'('`.
* Left parenthesis `'('` must go before the corresponding right parenthesis `')'`.
* `'*'` could be treated as a single right parenthesis `')'` or a single left parenthesis `'('` or an empty string `""`.


**Example 1:**

**Input:** s = "()"
**Output:** true

**Example 2:**

**Input:** s = "(*)"
**Output:** true

**Example 3:**

**Input:** s = "(*))"
**Output:** true


**Constraints:**

* `1 <= s.length <= 100`
* `s[i]` is `'('`, `')'` or `'*'`.
*/

use std::collections::VecDeque;

struct Solution;
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let mut left: Vec<(char, usize)> = Vec::new();
        let mut right: VecDeque<(char, usize)> = VecDeque::new();
        let mut astr: VecDeque<(char, usize)> = VecDeque::new();
        let n: usize = s.len();
        let s_chars: Vec<char> = s.chars().collect();

        for i in 0..n {
            let c: char = s_chars[i];

            if c == '(' {
                left.push((c, i));
            } else if c == ')' {
                if left.len() == 0 {
                    right.push_back((c, i));
                } else {
                    left.pop();
                }
            } else {
                astr.push_back((c, i));
            }
        }

        while left.len() != 0 && astr.len() != 0 {
            let left_top: (char, usize) = *left.last().unwrap();
            let astr_front: (char, usize) = *astr.front().unwrap();
            let astr_back: (char, usize) = *astr.back().unwrap();

            if left_top.1 < astr_front.1 {
                left.pop();
                astr.pop_front();
            } else if left_top.1 < astr_back.1 {
                left.pop();
                astr.pop_back();
            } else {
                break;
            }
        }

        while right.len() != 0 && astr.len() != 0 {
            let right_front: (char, usize) = *right.front().unwrap();
            let astr_front: (char, usize) = *astr.front().unwrap();
            let astr_back: (char, usize) = *astr.back().unwrap();

            if right_front.1 > astr_front.1 {
                right.pop_front();
                astr.pop_front();
            } else if right_front.1 > astr_back.1 {
                right.pop_front();
                astr.pop_back();
            } else {
                break;
            }
        }

        return left.is_empty() && right.is_empty();
    }
}

fn main() {
    let s_1: String = "()".to_string();
    let expected_1: bool = true;
    let output_1: bool = Solution::check_valid_string(s_1);
    println!("{:?}", output_1);
    assert!(output_1 == expected_1);

    let s_2: String = "(*)".to_string();
    let expected_2: bool = true;
    let output_2: bool = Solution::check_valid_string(s_2);
    println!("{:?}", output_2);
    assert!(output_2 == expected_2);

    let s_3: String = "(*))".to_string();
    let expected_3: bool = true;
    let output_3: bool = Solution::check_valid_string(s_3);
    println!("{:?}", output_3);
    assert!(output_3 == expected_3);
}
